nodebuilder/pruner/pruner.go

```diff
@@ -0,0 +1,239 @@
+package pruner
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-datastore/namespace"
+	"github.com/ipfs/go-datastore/query"
+	logging "github.com/ipfs/go-log/v2"
+
+	"github.com/celestiaorg/celestia-node/header"
+	"github.com/celestiaorg/celestia-node/share"
+	"github.com/celestiaorg/celestia-node/share/eds"
+)
+
+const (
+	dsPrefix = "/pruner/epoch/"
+)
+
+var log = logging.Logger("pruner")
+
+type StoragePruner struct {
+	cancel context.CancelFunc
+	cfg    Config
+
+	// TODO: Race?
+	oldestEpoch uint64
+
+	stripedLocks [256]sync.Mutex
+	activeEpochs map[uint64]struct{}
+
+	ds    datastore.Batching
+	store *eds.Store
+
+	done chan struct{}
+}
```

Comment:
general and only implementation comment: seeing the act of pruning via running the gc should/can be "background", it might be nice to put some sort of throttling/max worker/rate limit on how many blocks it tries to prune at any one run of the gc. Imagining a scenario where, for whatever reason, we get a big list of epochs to catch up on (not sure why, but lets just assume for some reason we do - maybe some resource pauses). Then a gc catchup run cant possibly ever degrade resources it touches (thinking deletes, badger, disk) by going ham in a loop when running through a big range.

might be nice to make this configurable too and test running prunung with resources constrained and observe out what amount of backpressure it could takes and what happens.

Reply:
I can see this potetnially happening on all the synced blocks that are synced for `sampling_window` prior